Chore(ebs): Add EBS By ID and Tag Experiment Docs (#470)
litmuschaos/litmus-docs: docs/ebs-loss-by-tag.md

---
id: ebs-loss-by-tag
title: EBS Loss By Tag Experiment Details
sidebar_label: EBS Loss By Tag
---
------

## Experiment Metadata

<table>
  <tr>
    <th> Type </th>
    <th>  Description  </th>
    <th> Tested K8s Platform </th>
  </tr>
  <tr>
    <td> Kube AWS </td>
    <td> EBS volume loss by Tag against specified application </td>
    <td> EKS </td>
  </tr>
</table>

## Prerequisites

- Ensure that Kubernetes Version > 1.15
- Ensure that the Litmus Chaos Operator is running by executing `kubectl get pods` in operator namespace (typically, `litmus`). If not, install from [here](https://docs.litmuschaos.io/docs/getstarted/#install-litmus)
- Ensure that the `ebs-loss-by-tag` experiment resource is available in the cluster by executing `kubectl get chaosexperiments` in the desired namespace If not, install from [here](https://hub.litmuschaos.io/api/chaos/master?file=charts/kube-aws/ebs-loss-by-tag/experiment.yaml)
- Ensure that you have sufficient AWS access to attach or detach an ebs volume from the instance.

> - Ensure the target volume to detach should not be the root volume from instance.

Reviewer: ```suggestion
- Ensure the target volume to detach should not be the root volume for the instance.
```

- Ensure to create a Kubernetes secret having the AWS access configuration(key) in the `CHAOS_NAMESPACE`. A sample secret file looks like:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cloud-secret
type: Opaque
stringData:
  cloud_config.yml: |-
    # Add the cloud AWS credentials respectively
    [default]
    aws_access_key_id = XXXXXXXXXXXXXXXXXXX
    aws_secret_access_key = XXXXXXXXXXXXXXX
```

- If you change the secret key name (from `cloud_config.yml`) please also update the `AWS_SHARED_CREDENTIALS_FILE` 
ENV value on `experiment.yaml`with the same name.

## Entry-Criteria

- Application pods are healthy before chaos injection
- EBS volume is attached to the instance.

## Exit-Criteria

-  Application pods are healthy post chaos injection 
-  EBS volume is attached to the instance.

## Details

> -  Causes chaos to disrupt state of infra resources ebs volume loss from node or ec2 instance for a certain chaos duration using a common volume tag.

Reviewer: ```suggestion
- Causes chaos to disrupt state of ebs volume by detaching it from the node/ec2 instance for a certain chaos duration using volume tags.
```

> -  In case of persistent volumes on top of EBS, the volumes can get self-attached and experiment skips the attachment procees in this case.

Reviewer: ```suggestion
-  In case of EBS persistent volumes, the volumes can get self-attached and experiment skips the re-attachment process
```

  resources: ["jobs"]
  verbs: ["create","list","get","delete","deletecollection"]
- apiGroups: ["litmuschaos.io"]
  resources: ["chaosengines","chaosexperiments","chaosresults"]
  verbs: ["create","list","get","patch","update"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: ebs-loss-by-tag-sa
  labels:
    name: ebs-loss-by-tag-sa
    app.kubernetes.io/part-of: litmus
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: ebs-loss-by-tag-sa
subjects:
- kind: ServiceAccount
  name: ebs-loss-by-tag-sa
  namespace: default
```

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`. It is an optional parameter for infra level experiment.
- Provide the auxiliary applications info (ns & labels) in `spec.auxiliaryAppInfo`
- Override the experiment tunables if desired in `experiments.spec.components.env`
- To understand the values to provided in a ChaosEngine specification, refer [ChaosEngine Concepts](chaosengine-concepts.md)

#### Supported Experiment Tunables

<table>
  <tr>
    <th> Variables </th>
    <th> Description </th>
    <th> Specify In ChaosEngine </th>
    <th> Notes </th>
  </tr>
  <tr> 

>      <td> EBS_VOLUME_ID </td>

Reviewer: Do we take VOL_ID in this tag-based experiment as well?

  jobCleanUpPolicy: 'delete'
  experiments:
    - name: ebs-loss-by-tag
      spec:
        components:
          env: 
            # set chaos duration (in sec) as desired
            - name: TOTAL_CHAOS_DURATION
              value: '30'

            - name: CHAOS_INTERVAL
              value: '30'

            # provide EBS volume tag attached to the given instance
            # it'll be in form of key:value (ex: 'team:devops')
            - name: EBS_VOLUME_TAG
              value: ''              
              
            # provide the region name of the instance
            - name: REGION
              value: ''
```

### Create the ChaosEngine Resource

- Create the ChaosEngine manifest prepared in the previous step to trigger the Chaos.

  `kubectl apply -f chaosengine.yml`

- If the chaos experiment is not executed, refer to the [troubleshooting](https://docs.litmuschaos.io/docs/faq-troubleshooting/) 
  section to identify the root cause and fix the issues.

### Watch Chaos progress

- View the status of the pods as they are subjected to ebs loss. 

  `watch -n 1 kubectl get pods -n <application-namespace>`
  
- Monitor the attachment status for ebs volume from AWS CLI.

>   `aws ec2 describe-volumes --volume-ids <vol-id>`

Reviewer: do we have any tag based filters?
